### ISSUE TITLE
Rubygems 2.2.2 uselessly hits the dependency API when installing a local gem file

### ISSUE THREAD

**Issue by indirect** [rubygems/rubygems#816]
The .gem file that I'm installing has the gemspec with the dependency information in it. The gem has no runtime dependencies. All 4 network requests are unnecessary to install the gem.

```
gem install pkg/bundler-1.5.3.gem --verbose
HEAD http://api.rubygems.org/api/v1/dependencies
302 Moved Temporarily
HEAD http://bundler.rubygems.org/api/v1/dependencies
200 OK
GET http://api.rubygems.org/api/v1/dependencies?gems=bundler
302 Moved Temporarily
GET http://bundler.rubygems.org/api/v1/dependencies?gems=bundler
200 OK
```

**Comment by copiousfreetime** [comment]
To clarify this issue:

If the gem to be installed is local, and all the dependencies are currently satisfied then make it a local install? Or do we want to only do this if if the gem being installed has zero dependencies?

I think this could be merged with #108

**Comment by sonalkr132** [comment]
As mentioned, two requests are originating from two different places:
*  [installer_set.add_always_install dependency](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/dependency_installer.rb#L474)

[`installer_set` should not ask](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/resolver/installer_set.rb#L156) `remote_set` to `find_all` dependencies of `req` if `req` has no dependencies. `req` is just a `Gem::Resolver::DependencyRequest` object, so it doesn't know about local `spec`(which would contain dependencies).

* [request_set.resolve installer_set](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/dependency_installer.rb#L484)

This can be solved if we [don't `prefetch` when `reqs` is `empty?`](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/resolver.rb#L167).

**Comment by deivid-rodriguez** [comment]
I added a fix for this at #3968.

> If the gem to be installed is local, and all the dependencies are currently satisfied then make it a local install? Or do we want to only do this if if the gem being installed has zero dependencies?

I assumed the latter option here in order to focus on fixing the issue without changing current behavior. But we could also consider to make installing `.gem` files conservative by default.
